### ISSUE TITLE
TksRegex modularization

### ISSUE THREAD

**Issue by bandrzejczak** [yegor256/takes#1]
I hava a concern. It's connected to the code below

```java
public final class App {
  public static void main(final String... args) {
    new TakesServer(
      new TksRegex()
        .with("/robots.txt", "")
        .with("/", new TkIndex())
        .with(
          "/xsl/.*", 
          new TkContentType(new TkClasspath(), "text/xsl")
        )
        .with(
          "/account", 
          new Take.Source() {
            @Override
            public Take take(final Request request) {
              return new TkAccount(users, request);
            }
          }
        )
        .with(
          "/balance/(?<user>[a-z]+)", 
          new TksRegex.Source() {
            @Override
            public Take take(final RqRegex request) {
              return new TkBalance(request);
            }
          }
        )
    ).listen();
  }
}
```

In this example you've configured only two resources (Account and Balance), but it took significant number of LOC. Have you thought of some way of grouping resource declarations into separate files?

**Comment by yegor256** [comment]
@bandrzejczak very good question. I though about this, but haven't documented it yet:

```java
public final class App {
  public static void main(final String... args) {
    new TakesServer(
      new TksRegex()
        .with("/robots.txt", "")
        .with("/", new TkIndex())
        .with("/users", new TksUsers())
    ).listen();
  }
}
```

Then, we create `TksUsers`:

```
public final class TksUsers implements Takes {
  @Override
  public Take take(final Request request) {
    return new TksRegex()
      .with(
        "/account", 
        new Take.Source() {
          @Override
          public Take take(final Request request) {
            return new TkAccount(users, request);
          }
        }
      ).take(request);
  }
}
```

What do you think?

**Comment by bandrzejczak** [comment]
It's quite nice, but you won't have deep nesting then. The question is whether it is a bad or a good thing. Take implementation returns Response from print method, while Takes implementation returns Take from take method. If you would like to have such deep nesting these are my thoughts on it: I don't really see a way to somehow merge Take and Takes interfaces. Even if I did, it would be stupid. Take could implement Response, so the Take could return either Response or Take. But this looks terribly stupid to me too. The only solution making sense is to have some class(interface?), that will be the link of the chain like below:

```java
public final class App {
  public static void main(final String... args) {
    new TakesServer(
      new TksRegex()
        .with("/robots.txt", "")
        .with("/", new TkIndex())
        .with("/users", new TksUsers())
    ).listen();
  }
}

public final class TksUsers implements ? {
  @Override
  public ? take() {
    return new TksRegex()
      .with("/users/account", new TksUserAccounts())
      .with("/users/group", new TksUserGroups())
      .take(request);
  }
}
```

But again, the real question is whether such deep nesting is useful at all, and necessary to implement at the beginning (maybe more people will see this as an advantage, and will propose so PR someday?). The question of implementation can be answered only after the first question.

**Comment by yegor256** [comment]
Here is how we can do deep nesting:

```java
public final class App {
  public static void main(final String... args) {
    new TakesServer(
      new TksRegex()
        .with("/robots.txt", "")
        .with("/", new TkIndex())
        .with("/users", new TksUsers())
    ).listen();
  }
}
```

Then, `TksUsers`:

```java
public final class TksUsers implements Takes {
  @Override
  public Take take(final Request request) {
    return new TksRegex()
      .with("/users/account", new TksUserAccounts())
      .with("/users/group", new TksUserGroups())
      .take(request);
  }
}
```

Then, `TksUserAccounts`:

```
public final class TksUserAccounts implements Takes {
  @Override
  public Take take(final Request request) {
    return new TksRegex()
      .with("/users/account/open", new TkUserAccountOpen())
      .with("/users/account/close", new TkUserAccountClose())
      .take(request);
  }
}
```

I don't see an issue here. Or I'm missing something?

**Comment by yegor256** [comment]
This is how `TksRegex` will look like:

```java
public final class TksRegex implements Takes {
  TksRegex with(String regex, Take take);
  TksRegex with(String regex, Takes takes);
  TksRegex with(String regex, String body);
}
```

**Comment by bandrzejczak** [comment]
It's hard to think about code, if you don't actually see the code ;) This looks totally fine to me. If you think, that the issue is resolved, please close it.

**Comment by yegor256** [comment]
@bandrzejczak would be great if you close it, since you opened it. thanks!

**Comment by bandrzejczak** [comment]
Sure thing ;) I'm new to the whole opensource contributing thing so I'm still learning. I'll let you know if I'll see another issue.
